### PR TITLE
Convert token controller to shared informers

### PR DESCRIPTION
**Pull request by hodovska**
Issue: https://github.com/kubernetes/kubernetes/issues/30016
@kubernetes/sig-apps-misc

### CODE REVIEW

**Comment by kargakis** [comment]
@deads2k @smarterclayton @liggitt any security concerns for the shared secrets cache?

File: pkg/client/cache/listers_core.go [hunk 1]
```diff
@@ -347,3 +347,43 @@ func (i *IndexerToNamespaceLister) Get(name string) (*v1.Namespace, error) {
 	}
 	return obj.(*v1.Namespace), nil
 }
+
+// StoreToSecretLister helps list secrets
+type StoreToSecretLister struct {
```
Comment (deads2k) [inline pkg/client/cache/listers_core.go]: These are all now generated.  Let's switch over the generated ones.

File: pkg/controller/serviceaccount/tokens_controller.go [hunk 2]
```diff
@@ -91,45 +88,23 @@ func NewTokensController(cl clientset.Interface, options TokensControllerOptions
...
-		&v1.ServiceAccount{},
-		options.ServiceAccountResync,
-		cache.ResourceEventHandlerFuncs{
-			AddFunc:    e.queueServiceAccountSync,
-			UpdateFunc: e.queueServiceAccountUpdateSync,
-			DeleteFunc: e.queueServiceAccountSync,
-		},
-	)
-
-	tokenSelector := fields.SelectorFromSet(map[string]string{api.SecretTypeField: string(v1.SecretTypeServiceAccountToken)})
```
Comment (deads2k) [inline pkg/controller/serviceaccount/tokens_controller.go]: I'm ok with losing this field selector.  Seems like better than half of all secrets are SA secrets and this lets us share the cache.  I'd add checks before using the values though.

@liggitt call out, but with appropriate checking elsewhere, given the overall usage secrets, I suspect that this is worth it.

Comment (liggitt) [inline pkg/controller/serviceaccount/tokens_controller.go]: definitely needs to filter on secret type in the event handlers

File: pkg/controller/serviceaccount/tokens_controller.go [hunk 3]
```diff
@@ -431,7 +403,7 @@ func (e *TokensController) ensureReferencedToken(serviceAccount *v1.ServiceAccou
 	}
 	// Manually add the new token to the cache store.
 	// This prevents the service account update (below) triggering another token creation, if the referenced token couldn't be found in the store
-	e.secrets.Add(createdToken)
+	e.secretLister.Indexer.Add(createdToken)
```
Comment (liggitt) [inline pkg/controller/serviceaccount/tokens_controller.go]: I question whether this is ok now that the cache is shared.

Comment (hodovska) [inline pkg/controller/serviceaccount/tokens_controller.go]: @liggitt Could we now drop this prevention or should we sync/wait?

Comment (liggitt) [inline pkg/controller/serviceaccount/tokens_controller.go]: > @liggitt Could we now drop this prevention or should we sync/wait?

no, dropping the prevention exposes the controller to racing caches between the two object types. there's not a great condition to wait on...

**Comment by ncdc** [comment]
FYI I am working on a giant PR to replace the majority of standalone informers with shared ones. Would you mind if I incorporated your work into that one, and closed this PR?

**Comment by ncdc** [comment]
@hodovska I just opened #40097 which includes changes for the tokens controller

**Comment by hodovska** [comment]
@ncdc I don't mind, thank you for letting me know :)

**Pull request closed without merging**